src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -0,0 +1,205 @@
+import static com.epam.esm.exceptions.Messages.CERTIFICATE_ALREADY_EXISTS;
+import static com.epam.esm.exceptions.Messages.CERTIFICATE_WITH_ID_NOT_FOUND;
+
+import com.epam.esm.Dto.Errors.ErrorDTO;
+import com.epam.esm.exceptions.CertificateNotFoundException;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
+import com.epam.esm.repository.CertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.utils.CertificateSpecification;
+import jakarta.transaction.Transactional;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class CertificateService {
+
+    private final CertificateRepository certificateRepository;
+    private final TagRepository tagRepository;
+
+    @Autowired
+    public CertificateService(CertificateRepository certificateRepository, TagRepository tagRepository) {
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+    }
+
+    public ResponseEntity<?> saveGiftCertificate(@NonNull GiftCertificate giftCertificate, List<Long> tagIdsList) {
+
+        Optional<ResponseEntity<ErrorDTO>> requestValidationMessage = validateCertificateRequest(giftCertificate, tagIdsList);
+        if (requestValidationMessage.isPresent()) return requestValidationMessage.get();
+
+        tagIdsList = tagIdsList.stream().distinct().collect(Collectors.toList());
+        Optional<GiftCertificate> tryToFindCertificate = certificateRepository.findByName(giftCertificate.getName());
+
+        List<Tag> tags = tagRepository.findAllById(tagIdsList);
+        List<Tag> uniqueTags = new ArrayList<>(new HashSet<>(tags));
+        giftCertificate.setTags(uniqueTags);
+
+        if (tryToFindCertificate.isEmpty()) {
+            GiftCertificate savedGiftCertificate = certificateRepository.save(giftCertificate);
+                return new ResponseEntity<>(savedGiftCertificate, HttpStatus.CREATED);
+
+        } else {
+            GiftCertificate foundCertificate = tryToFindCertificate.get();
+            Long idFound = foundCertificate.getId();
+            ErrorDTO errorResponse = new ErrorDTO(CERTIFICATE_ALREADY_EXISTS.formatted(idFound), CERTIFICATE_FOUND);
+            return ResponseEntity.status(HttpStatus.FOUND).body(errorResponse);
+        }
+    }
+
+    public ResponseEntity<?> getGiftCertificateById(@NonNull Long giftCertificateId) {
+
+        if (certificateRepository.existsById(giftCertificateId)) {
```

Comment:
You are making two calls to DB instead of one (first "exstsById" than "findById"). Just do "findById"  immediately and check Optional for emptiness.

Reply:
Done